File: Algorithm/Heuristic/atcoder/GreedyAlgorithm/A49/A49.py
import sys

input = sys.stdin.read
data = input().split()

T = int(data[0])
triplets = [tuple(map(lambda x: int(x) - 1, data[i:i+3])) for i in range(1, len(data), 3)]

X = [0] * 20
zero_count = 20
result = []

for p, q, r in triplets:
    indices = [p, q, r]
    gain_a = 0
    gain_b = 0

    # 操作A (+1)
    for idx in indices:
        if X[idx] == 0:
            gain_a -= 1
        if X[idx] + 1 == 0:
            gain_a += 1

    # 操作B (-1)
    for idx in indices:
        if X[idx] == 0:
            gain_b -= 1
        if X[idx] - 1 == 0:
            gain_b += 1

    if zero_count + gain_a >= zero_count + gain_b:
        # A操作を実行
        result.append("A")
        for idx in indices:
            if X[idx] == 0:
                zero_count -= 1
            X[idx] += 1
            if X[idx] == 0:
                zero_count += 1
    else:
        # B操作を実行
        result.append("B")
        for idx in indices:
            if X[idx] == 0:
                zero_count -= 1
            X[idx] -= 1
            if X[idx] == 0:
                zero_count += 1

print("\n".join(result))

File: Algorithm/Heuristic/atcoder/GreedyAlgorithm/A49/A49.ts
import * as readline from 'readline';

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
});

const input: string[] = [];

rl.on('line', (line) => {
    input.push(line.trim());
}).on('close', () => {
    const T = Number(input[0]);
    const ops: [number, number, number][] = input.slice(1).map((line) => {
        const [p, q, r] = line.split(' ').map((v) => Number(v) - 1); // 0-indexed
        return [p, q, r];
    });

    const X = new Int8Array(20); // 配列X（20要素）
    let zeroCount = 20; // 現在の0の個数
    const result: string[] = [];

    for (let i = 0; i < T; i++) {
        const [p, q, r] = ops[i];
        const indices = [p, q, r];

        // A操作（+1）によるスコア変化
        let gainA = 0;
        for (const idx of indices) {
            if (X[idx] === 0) gainA--;
            if (X[idx] + 1 === 0) gainA++;
        }

        // B操作（-1）によるスコア変化
        let gainB = 0;
        for (const idx of indices) {
            if (X[idx] === 0) gainB--;
            if (X[idx] - 1 === 0) gainB++;
        }

        if (zeroCount + gainA >= zeroCount + gainB) {
            // 操作Aを選択
            result.push('A');
            for (const idx of indices) {
                if (X[idx] === 0) zeroCount--;
                X[idx]++;
                if (X[idx] === 0) zeroCount++;
            }
        } else {
            // 操作Bを選択
            result.push('B');
            for (const idx of indices) {
                if (X[idx] === 0) zeroCount--;
                X[idx]--;
                if (X[idx] === 0) zeroCount++;
            }
        }
    }

    console.log(result.join('\n'));
});
